draft-specifications/Model.md

```diff
@@ -0,0 +1,326 @@
+>
+> We achieved consensus to name the type "List" at the [November 14, 2023](https://github.com/ocapn/ocapn/issues/94) meeting.
+
+## Struct (JSON)
+
+A struct with unique, unordered string keys and values of heterogeneous type.
+
+> - **Guile**: `make-tbd-hash` a hash of undecided type
+> - **JavaScript**: `{}`
+> - **Python**: `{}`
+>
+> For the purposes of surviving a round trip, the order of appearance of
+> entries in the struct must not be important for determining equivalence.
+> A struct representation concretely using one key order may validly
+> round-trip into a struct representation using another key order.
+> However, we encourage concrete representations to use some canonical key
+> order, though this is not a concern of the abstract syntax and data model.
+>
+> A JavaScript object that owns any non-string keys (numbers or symbols) or
+> with a prototype other than `Object.prototype cannot be passed as an OCapN
+> struct as all such objects are reserved for concrete representations of OCapN
+> types.
+> The key `Symbol.for('passStyle')` is special and indicates the kind of OCapN
+> value the object represents.
+
+## Tagged
+
+A value with a string tag.
+
+> OCapN provides a small number of container types with minimal semantics.
+> Tags allow protocols to emerge from values annotated to indicate a richer
+> variety of semantics.
+> Concretely, a list might be differentiated from or hinted to be an ordered
+> list, sorted list, or a set of unique values.
+>
+> - **Guile**: currently a Guile record labeled `'(syrec ...)` (**imported**)
+> - **JavaScript**: an object with the key `Symbol.for('passStyle')` and value
+>   `tagged`, with a string tagName and a single `"payload"` field carrying the tagged value.
+>   ```js
+>   {
+>     [Symbol.for('passStyle')]: 'tagged',
+>     [Symbol.toStringTag]: tagName,
+>     payload
+>   }
+>   ```
+> - **Python**: `Tagged(tag, value)` where `tag` is a `string` and `Tagged` is
+>   imported from `ocapn`.
+>
+> OCapN tagged data have a single value (are not variadic) [November 14, 2023](https://github.com/ocapn/ocapn/issues/94)
+>
+> What are tagged values for? https://github.com/ocapn/ocapn/issues/52
+
+## Remotable
+
+A value that can receive messages.
+
+Remotable values can be either a local value or a remote value that forwards
+messages to a local value.
+The latter is a presence: a remotable that forwards messages to the peer where the
+message can be delivered to its corresponding local remotable.
+
+Comparing the identities of remotable values is pass-invariant, meaning that
+comparing the identities of any pair of remotable values will produce the same
+result on one peer if passed over OCapN to another peer and compared there,
+regardless of whether either value is a presence on either peer.
+
+Although Remotables and Promises are both Capabilities for delivering messages,
+Promises do not support pass-invariant identity comparison.
+
+> - **Guile**: a function
```

Comment:
s/function/procedure/ but idk a callable actor is not a core concept in Guile, it's a Goblins thing.

Reply:
Are all Guile procedures eligible targets for message delivery or do they need to be marked with a Guile thing?

Reply:
So I guess there's a bit of a blurring of the lines between Guile and Goblins. In terms of references, Goblins has two object/actor references:

`<local-object-refr>` this is a reference which refers to a guile procedure (the object's/actor's behaviour) that was returned by spawning the actor's constructor.
`<remote-object-refr>` which is a reference to an actor on a remote system.

In addition to those, you can also send messages to promises, there are two equivalent promise references:
- `<local-promise-refr>`
- `<remote-promise-refr>`

When Goblin's CapTP layer gets a message to a `<local-object-refr>` it'll queue that message on the vat's message queue so that the vat invokes that procedure in turn.

Reply:
Endo/Agoric have analogous local and remote targets, though we just use a JavaScript promise for promises regardless of remote vs local. We call a remote target a “presence” which is just an opaque object that we can send messages to.

Follow-up question: Does Goblins oblige you to wrap a procedure to construct a `<local-object-refr>` in order for it to be able to receive deliveries? In JavaScript, we mark objects with `makeExo(interfaceName, interfaceGuard, methods)`, which is a user mode utility for generating an object that we recognize as an eligible target, as opposed to arbitrary JavaScript objects. This distinction prevents the JavaScript API from implicitly leaking into protocols above CapTP. I imagine this is less necessary for Goblins, but your answer may come to bear on the accuracy of this non-normative example.